src/api/java/Context.java

```diff
@@ -548,28 +549,28 @@ public FuncDecl mkConstDecl(String name, Sort range)
      * @see #mkFuncDecl(String,Sort,Sort)
      * @see #mkFuncDecl(String,Sort[],Sort)
      **/
-    public FuncDecl mkFreshConstDecl(String prefix, Sort range)
+    public <R extends Sort> FuncDecl<R> mkFreshConstDecl(String prefix, R range)
 
     {
         checkContextMatch(range);
-        return new FuncDecl(this, prefix, null, range);
+        return new FuncDecl<>(this, prefix, null, range);
     }
 
     /**
      * Creates a new bound variable. 
      * @param index The de-Bruijn index of the variable 
      * @param ty The sort of the variable
      **/
-    public Expr mkBound(int index, Sort ty)
+    public <R extends Sort> Expr<R> mkBound(int index, R ty)
     {
-        return Expr.create(this,
+        return (Expr<R>) Expr.create(this,
                 Native.mkBound(nCtx(), index, ty.getNativeObject()));
     }
 
     /**
      * Create a quantifier pattern.
      **/
-    public Pattern mkPattern(Expr... terms)
+    public Pattern mkPattern(Expr<?>... terms)
```

Comment:
Is there anything we can say consistently about the types of the terms in a pattern? Not sure how this is restricted in z3.